File: languages/javascript/hron.js
/*!
 * HRON v0.8.0
 * A Javascript module for parsing and serializing HRON
 * https://github.com/mrange/hron
 * Microsoft Public License (Ms-PL)
 * by Daniel Brännström
 */

(function(exports) {

	'use strict';

	////////////////////////////////////////////////////////////////////
	// Utilities
	////////////////////////////////////////////////////////////////////

	function isOfArrayType(o) {
		return Object.prototype.toString.call(o) === '[object Array]';
	}

	// ref: http://gomakethings.com/vanilla-javascript-version-of-jquery-extend/
	function extend (defaults, options) {
	    var extended = {};
	    var prop;
	    for (prop in defaults) {
	        if (Object.prototype.hasOwnProperty.call(defaults, prop)) {
	            extended[prop] = defaults[prop];
	        }
	    }
	    for (prop in options) {
	        if (Object.prototype.hasOwnProperty.call(options, prop)) {
	            extended[prop] = options[prop];
	        }
	    }
	    return extended;
	}

	////////////////////////////////////////////////////////////////////
	// Deserialization (parsing)
	////////////////////////////////////////////////////////////////////

	var reCommentLine = new RegExp("^(\\s*)#(.*)");
	var reEmptyLine = new RegExp("^(\\s*?)\\r?$");
	var rePreprocessorLine = new RegExp("^!(.*)");

	function RegExCache(regexFactory) {
		var cacheSize = 10;
		var cache = [];
		for(var indent = 0; indent < cacheSize; ++indent) {
			cache.push(regexFactory(indent));
		}
		this.get = function(indent) {
			return indent < cacheSize ? cache[indent] : regexFactory(indent);
		};
	}

	var reNonEmptyLineCache = new RegExCache(function(indent) {
		return new RegExp("^\\t{" + indent + "}(.*)");
	});

	var reValueDefCache = new RegExCache(function(indent) {
		return new RegExp("^\\t{" + indent + "}=(.*)");
	});

	var reObjectDefCache = new RegExCache(function(indent) {
		return new RegExp("^\\t{" + indent + "}@(.*)");
	});

	function deserializePreprocessors(state) {
		/* jshint -W084 */
		var match;
		while(match = state.currentLine().match(rePreprocessorLine)) {
			state.skipLine("PreProcessor", match[1]);
		}
		/* jshint +W084 */
	}

	function deserializeValueLines(state) {
		var reNonEmptyLine = reNonEmptyLineCache.get(state.currentIndent);
		var match;
		var stop = false;
		var result = [];
		while(!stop && !state.eos()) {
			match = state.currentLine().match(reNonEmptyLine); 
			if (match) {
				result.push(match[1]);
				state.skipLine("ContentLine", match[1]);
				continue;
			}	

			match = state.currentLine().match(reCommentLine);
			if (match) {
				state.skipLine("CommentLine", match[1].length + "," + match[2]);
				continue;
			}

			match = state.currentLine().match(reEmptyLine);
			if (match) {
				state.skipLine("EmptyLine", match[1]);
				continue;
			}

			stop = true;
		} 

		return result.join("\n");
	}

	function tryDeserializeValue(state) {
		var re = reValueDefCache.get(state.currentIndent);
		var match = state.currentLine().match(re);
		var result;
		if (match) {
			state.skipLine("Value_Begin", match[1]);
			result = { key: match[1] };
			++state.currentIndent;
			result.value = deserializeValueLines(state);
			--state.currentIndent;
			state.logAction("Value_End", match[1]);
		}

		return result;
	}

	function tryDeserializeObject(state) {
		var re = reObjectDefCache.get(state.currentIndent);
		var match = state.currentLine().match(re);
		var result;
		if (match) {
			state.skipLine("Object_Begin", match[1]);
			result = { key: match[1] };
			++state.currentIndent;
			state.objectStack.push({});
			deserializeMembers(state);
			result.value = state.currentObject();
			state.objectStack.pop();
			--state.currentIndent;
			state.logAction("Object_End", match[1]);
		}

		return result;
	}

	function addPropertyToCurrentObject(state, name, value) {
		var o = state.currentObject();
		if (isOfArrayType(o[name])) {
			o[name].push(value);
		}
		else if (o[name]) {
			o[name] = [o[name], value];
		}
		else {
			o[name] = value;
		}
	}

	function deserializeMembers(state) {
		var stop = false;
		while(!stop && !state.eos()) {
			var value = tryDeserializeValue(state);
			if (value) {
				addPropertyToCurrentObject(state, value.key, value.value);
				continue;
			}		

			var object = tryDeserializeObject(state);
			if (object) {
				addPropertyToCurrentObject(state, object.key, object.value);
				continue;
			}

			var match = state.currentLine().match(reCommentLine); 
			if (match) {
				state.skipLine("Comment", match[1].length + "," + match[2]);
				continue;
			}

			match = state.currentLine().match(reEmptyLine);
			if (match) {
				state.skipLine("Empty", match[1]);
				continue;
			}		

			stop = true;
		}
	}

	function DeserializationState(text, options) {
		var defaults = { actionLog : null };
		var settings = extend(defaults, options);
		var lines = text.replace(/\n$/,"").split("\n"); // use replace to avoid empty line at the end
		var lineIndex = 0;
		this.currentIndent = 0;	
		this.objectStack = [{}];
		this.logAction = function(action, info) {
			if (action && settings.actionLog) {
				settings.actionLog.push(action + ":" + (info || ""));
			}
		};
		this.currentObject = function() {
			return this.objectStack[this.objectStack.length-1];
		};
		this.currentLine = function() {
			return lines[lineIndex];
		};
		this.skipLine = function(action, logtext) {
			this.logAction(action, logtext);
			++lineIndex;
		};
		this.eos = function() {
			return lineIndex >= lines.length;	
		};	
	}

	function deserialize(text, options) {
		var state = new DeserializationState(text, options);
		deserializePreprocessors(state);
		deserializeMembers(state);
		return state.currentObject();
	}

	////////////////////////////////////////////////////////////////////
	// Serialization
	////////////////////////////////////////////////////////////////////

	function SerializationState() {
		this.indent = "";
		this.result = [];
		this.push = function(str) {
			this.result.push(this.indent + str);
		};
		this.incIndent = function() {
			this.indent += "\t";
		};
		this.decIndent = function() {
			this.indent = this.indent.substring(0, this.indent.length-1);
		};
	}

	function serializeMembers(state, object) {
		var keys = [];
		var key;
		for (key in object) {
			keys.push(key);
		}
		keys.sort();

		var len = keys.length;
		for (var i=0; i<len; ++i) { 
			key = keys[i];
			var value = object[key];
			if (isOfArrayType(value)) {
				var len_ = value.length;
				for (var j = 0; j < len_; ++j) {
					serializeInstance(state, key, value[j]);
				}
			}
			else {
				serializeInstance(state, key, value);
			}
		}
	}

	function serializeInstance(state, name, instance) {		
		var symbol = typeof instance === 'object' ? '@' : '='; 
		state.push(symbol + name);
		state.incIndent();
		
		switch(typeof instance) {
			case 'string':
				var lines = instance.split('\n');
				var len = lines.length;
				for (var i = 0; i < len; ++i) {
					state.push(lines[i]);
				}
				break;
			case 'object':			
				serializeMembers(state, instance);				
				break;
			default:
				if (instance) {
					state.push(String(instance));
				}
				break;
		}
		
		state.decIndent();				
	}

	function serialize(object) {
		var state = new SerializationState();
		serializeMembers(state, object);
		return state.result.join('\n');
	}


	////////////////////////////////////////////////////////////////////
	// Exports
	////////////////////////////////////////////////////////////////////

	exports.parse = deserialize;
	exports.serialize = serialize;

})(typeof exports === 'undefined'? this.hron = {} : exports);

File: languages/python/hron.py
####################################################################
# hron parsing and serialization library
# 
# Daniel Brännström, 2015
####################################################################

import re
import collections

####################################################################
# parsing (deserialization)
####################################################################

class _RegexCache:
    def __init__(self, functor):
        self.functor = functor
        self.cache = [ re.compile(functor(i)) for i in range(0,10) ]
    def get(self, index):
        if index < len(self.cache):
            return self.cache[index]
        else:
            return re.compile(self.functor(index))

class _Patterns:
    preprocessor = re.compile("^!(.*)")
    commentLine = re.compile("^(\\s*)#(.*)")
    emptyLine = re.compile("^(\\s*?)\\r?$")
    nonEmptyLineCache = _RegexCache(lambda indent: "^\\t{" + str(indent) + "}(.*)")
    valueDefCache = _RegexCache(lambda indent: "^\\t{" + str(indent) + "}=(.*)")
    objectDefCache = _RegexCache(lambda indent: "^\\t{" + str(indent) + "}@(.*)")

class _Dynamic:
    pass;

class _DeserializationState:
    def __init__(self, text):
        self.lines = text
        self.currentIndent = 0
        self.actionLog = None
        self.objectStack = [_Dynamic()]
        self.index = 0
    @property
    def currentLine(self):
        return self.lines[self.index]
    @property
    def currentObject(self):
        return self.objectStack[-1]
    @property
    def eos(self):
        return self.index >= len(self.lines)
    def log(self, action, info):
        if action and self.actionLog != None:
            self.actionLog.append(action + ":" + info)
    def skipLine(self, action, info):
        self.log(action, info)
        self.index += 1
    def enableLogging(self):
        self.actionLog = []

def _deserializePreprocessors(state):
    match = _Patterns.preprocessor.search(state.currentLine)
    while match: 
        state.skipLine("PreProcessor", match.group(1))
        match = _Patterns.preprocessor.search(state.currentLine)

def _deserializeValueLines(state):
    reNonEmptyLine = _Patterns.nonEmptyLineCache.get(state.currentIndent)
    stop = False
    result = []
    while(not(stop) and not(state.eos)):
        match = reNonEmptyLine.search(state.currentLine)
        if match:
            result.append(match.group(1))
            state.skipLine("ContentLine", match.group(1))
            continue

        match = _Patterns.commentLine.search(state.currentLine)
        if match: 
            state.skipLine("CommentLine", str(len(match.group(1))) + "," + match.group(2))
            continue

        match = _Patterns.emptyLine.search(state.currentLine)
        if match:
            state.skipLine("EmptyLine", match.group(1))
            continue
        
        stop = True    

    return "\n".join(result)

def _tryDeserializeValue(state):
    re = _Patterns.valueDefCache.get(state.currentIndent)
    match = re.search(state.currentLine)
    result = None
    if match:
        state.skipLine("Value_Begin", match.group(1))
        key = match.group(1)
        state.currentIndent += 1
        value = _deserializeValueLines(state)
        state.currentIndent -= 1
        state.log("Value_End", match.group(1))
        result = (key, value)    

    return result

def _tryDeserializeObject(state):
    re = _Patterns.objectDefCache.get(state.currentIndent)
    match = re.search(state.currentLine)
    result = None
    if match:
        state.skipLine("Object_Begin", match.group(1))
        key = match.group(1)
        state.currentIndent += 1
        state.objectStack.append(_Dynamic())
        _deserializeMembers(state)
        value = state.objectStack.pop();
        state.currentIndent -= 1
        state.log("Object_End", match.group(1))
        result = (key, value)

    return result

def _addPropertyToCurrentObject(state, name, value):
    o = state.currentObject
    if hasattr(o, name):
        a = getattr(o, name)
        if type(a) is list:
            a.append(value)
        else:
            setattr(o, name, [a, value])
    else:
        setattr(o, name, value)

def _deserializeMembers(state):
    stop = False
    while(not(stop) and not(state.eos)):
        value = _tryDeserializeValue(state)
        if value:
            _addPropertyToCurrentObject(state, value[0], value[1])
            continue

        object = _tryDeserializeObject(state)
        if object:
            _addPropertyToCurrentObject(state, object[0], object[1])
            continue

        match = _Patterns.commentLine.search(state.currentLine)
        if match:
            state.skipLine("Comment", str(len(match.group(1))) + "," + str(match.group(2)))
            continue

        match = _Patterns.emptyLine.search(state.currentLine)
        if match:
            state.skipLine("Empty", match.group(1))
            continue

        stop = True            

def _deserialize(arg):
    state = arg if isinstance(arg, _DeserializationState) else _DeserializationState(arg)
    _deserializePreprocessors(state)
    _deserializeMembers(state)
    return state.currentObject

def parse(arg):
    return _deserialize(arg)

####################################################################
# serialization
####################################################################

class _SerializationState:
    def __init__(self):
        self.indent = ""
        self.result = []

    def push(self, str):
        self.result.append(self.indent + str);
    
    def incIndent(self):
        self.indent += "\t";
    
    def decIndent(self):
        self.indent = self.indent[0:-1];
    
def _getKeyValues(object):
    if isinstance(object, dict):
        return list(object.items())
    else:
        return [(a,getattr(object,a)) for a in dir(object) if not a.startswith('__') and not callable(getattr(object,a))]

def _serializeMembers(state, object):
    keyValues = _getKeyValues(object)
    keyValues.sort(key=lambda kv: kv[0])

    for key, value in keyValues:
        if isinstance(value, collections.Iterable) and not(isinstance(value, (str,dict))):
            for elem in iter(value):
                _serializeInstance(state, key, elem)
        else:
            _serializeInstance(state, key, value)

def _serializeInstance(state, name, instance):
    symbol = '=' if isinstance(instance, (str, int, float)) else '@' 
    state.push(symbol + name)
    state.incIndent()

    if isinstance(instance, str):
        for line in instance.split("\n"):
            state.push(line);
    elif isinstance(instance, (int, float)): 
        state.push(str(instance));                
    else:   
        _serializeMembers(state, instance);        
        
    state.decIndent();                

def serialize(object):
    state = _SerializationState();
    _serializeMembers(state, object);
    return "\n".join(state.result);

